designs/dynamic-bundle.md

```diff
@@ -0,0 +1,267 @@
+Dynamic Bundle
+===================
+
+## Overview
+Kubebuilder currently has an implementation of a `Bundle` which allows multiple plugins to be *bundled* together. The execution of the plugins within a `Bundle` follows a sequential order based on their definition. 
+
+An example of the current method:
+
+```go
+gov3Bundle, _ := plugin.NewBundle(golang.DefaultNameQualifier, plugin.Version{Number: 3},
+	kustomizecommonv1.Plugin{},
+	golangv3.Plugin{},
+)
+```
+
+In the above example a plugin `Bundle` is created that ensures that the `kustomize/v1` plugin is run before the `go/v3` plugin.
+ 
+With the addition of Phase 2 Plugins on the way, users using the `--plugins` flag to utilize external plugins will not get the benefits of using some default functionality that previous Kubebuilder users using the default `go/v3` plugin gained via the `kustomize/v1` plugin.
+
+## Proposal
+The proposed solution is to create a new form of `Bundle` that is more dynamic and can be used to create a default bundle that contains sensible default plugins that can be used with all plugins. This could be overriden by a command line flag.
+
+## User Stories
+- As a Kubebuilder maintainer I would like to have Kubebuilder implement sane default plugins that always run before and/or after plugins, specifying this once and it applying to all uses
+
+- As a Kubebuilder user I expect Kubebuilder to have default functionality to make my operator development easier
+
+- As a Kubebuilder user I would like to be able to override the default Kubebuilder plugin chain, scaffolding only exactly what I feel I need
+
+## Goals
+- `kubebuilder` is able to define sane defaults for plugins that should always be run before and/or after user specified plugins
+
+- A `Bundle` no longer has to be created for each plugin to implement sane defaults
+
+## Examples
+Context: Kubebuilder defines that the `kustomize/v1` plugin should always run before any other plugin. Kubebuilder also defines the default plugin to be run if a user does not specify plugins via the `--plugins` flag is `go/v3`
+
+- `kubebuilder init`
+	- Should scaffold using the `kustomize/v1` plugin and the `go/v3` plugin
+- `kubebuilder init --plugins other.plugin/v3`
+	- Should scaffold using the `kustomize/v1` plugin and the `other.plugin/v3` plugin
+- `kubebuilder init --plugins other.plugin/v3 --override-default-plugin-chain`
+	- Should scaffold using only the `other.plugin/v3` plugin
```

Comment:
In this way, we are here proposing that kubebuilder init should only scaffold the kustomize and then we would no longer have the default plugins used by the CLI?

See that we define the default plugin that should be used by Kubebuilder/SDK or any project that uses Kubebuilder CLI Lib so that when a user runs kubebuilder init, that means scaffold the default plugin is. Today it goes/v3. 

What are the advantages for kubebuilder users?
How that would make the UX and usability better?

Reply:
Hi @everettraven:

Following some coments inline: 

> This is rather proposing that a dynamic bundle be implemented so that there can be default plugins that run no matter what plugins a user passes in. These examples are just meant to show what it would be like if Kubebuilder implemented this functionality.

We have the default plugin option already and it has been used
Also, we can overwrite the default plugin by --plugins informed via the flag

> In these examples, the context is that Kubebuilder wants to make sure the kustomize plugin is always run before any other plugin.

Why would we force run kustomize in all cases?
See that API is valid for any plugin/CLI and kustomize does not fit in any scenario. 
As shared before, we indeed have a proposal in place to config the projects using the config-gen approach instead of kustomize. Also, we have people that do not use kustomize at all, see: https://github.com/kubernetes-sigs/kubebuilder/pull/2350

So, if the main goal of this proposal is to force kustomize ( as it seems to be ) I do not think that we can go along with it.